File: python/Q347_top_k_frequent_elements.py
#
# @lc app=leetcode id=347 lang=python3
#
# [347] Top K Frequent Elements
#


# @lc code=start
import heapq


class Solution:
    def sorting(self, nums: list[int], k: int) -> list[int]:
        frequent_count: dict[int, int] = dict()

        for n in nums:
            frequent_count[n] = frequent_count.get(n, 0) + 1

        result: list[tuple[int, int]] = sorted(
            frequent_count.items(), key=lambda x: x[1], reverse=True
        )

        return [n for n, c in result[:k]]

    def min_heap(self, nums: list[int], k: int) -> list[int]:
        frequent_count: dict[int, int] = dict()

        for n in nums:
            frequent_count[n] = frequent_count.get(n, 0) + 1

        heap: list[tuple[int, int]] = []

        for n, c in frequent_count.items():
            heapq.heappush(heap, (-c, n))

        result: list[int] = []
        for _ in range(k):
            if not heap:
                break

            result.append(heapq.heappop(heap)[1])

        return result

    def bucket_sort(self, nums: list[int], k: int) -> list[int]:
        frequent_count: dict[int, int] = dict()

        for n in nums:
            frequent_count[n] = frequent_count.get(n, 0) + 1

        buckets: list[list[int]] = [[] for _ in range(len(nums) + 1)]

        for n, c in frequent_count.items():
            buckets[c].append(n)

        result: list[int] = []

        for i in range(len(buckets) - 1, -1, -1):
            while len(result) < k and buckets[i]:
                result.append(buckets[i].pop())

        return result

    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        return self.bucket_sort(nums, k)


# @lc code=end

File: rust/src/leetcode/Q347_top_k_frequent_elements.rs
/*
 * @lc app=leetcode id=347 lang=rust
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
use std::collections::BinaryHeap;
use std::collections::HashMap;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut counts: HashMap<i32, i32> = HashMap::new();

        for n in nums {
            *counts.entry(n).or_insert(0) += 1;
        }

        let mut frequency_heap: BinaryHeap<(i32, i32)> = BinaryHeap::new();

        for (num, frequency) in counts {
            frequency_heap.push((frequency, num));
        }

        let mut result: Vec<i32> = vec![];

        for _ in 0..k {
            if let Some((_, num)) = frequency_heap.pop() {
                result.push(num);
            }
        }

        result
    }
}
// @lc code=end
